### PR TITLE
Custom engines per template module

### PR DESCRIPTION
**Pull request by amarraja**

### CODE REVIEW

**Comment by amarraja** [comment]
@josevalim some of the context got lost when moving this to a PR so will comment here. 

Re: removing the `engine/0` and corresponding `defoverridable` statements. I put these in so this test and the similar one I added would pass:

```elixir
test "engines/0" do
  assert is_map Template.engines
end
```

If I remove them, `MyViewEngine.engines()` would just return the default + global engines no?

File: lib/phoenix/template.ex [hunk 1]
```diff
@@ -177,6 +179,10 @@ defmodule Phoenix.Template do
         template_not_found(template, Map.put(assigns, :__phx_template_not_found__, __MODULE__))
       end
 
+      def engines() do
```
[the remark this answers is not in the document]
Comment (amarraja) [reply lib/phoenix/template.ex]: Aah, got you! I agree - I think it was some OOP leaking out of my fingers, turns out it didn't really do what I thought it did anyway

File: lib/phoenix/template.ex [hunk 2]
```diff
@@ -321,8 +335,14 @@ defmodule Phoenix.Template do
   Used by Phoenix to check if a given root path requires recompilation.
   """
   @spec hash(root, pattern :: String.t) :: binary
-  def hash(root, pattern \\ @default_pattern) do
-    find_all(root, pattern)
+  def hash(root, pattern \\ @default_pattern)
```
[the remark this answers is not in the document]
Comment (amarraja) [reply lib/phoenix/template.ex]: Not sure why I didn't think of that. Much cleaner! Thanks!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by amarraja** [comment]
Thanks for the help @josevalim, was a small change but I learned a lot!

**Pull request merged**